infer/environment.cc

```diff
@@ -1715,6 +1726,8 @@ Environment::processBinding(core::Context ctx, const cfg::CFG &inWhat, cfg::Bind
                                 e.setHeader("Incompatible assignment to variable declared via `{}`: `{}` is not a "
                                             "subtype of `{}`",
                                             "let", tp.type.show(ctx), cur.type.show(ctx));
+                                core::TypeErrorDiagnostics::explainTypeMismatch(ctx, e, errorDetailsCollector, tp.type,
```

Comment:
I wasn't able to come up with a test case for this either.

Reply:
```ruby
# typed: true

xs = T.let([0], [Integer])
1.times do
  xs = [""]
end
```

You don't need to write a test case for this if you don't want to, it should virtually always be covered by the other tests (from a subtyping perspective there's nothing particularly special here).